fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -87,42 +90,149 @@ void removeTrailingChar(std::string& str, char c) {
 	}
 }
 
+constexpr double LAST_5MIN_PENALTY_WEIGHT = 1.0;
+constexpr double LAST_15MIN_PENALTY_WEIGHT = 0.90;
+constexpr double LAST_30MIN_PENALTY_WEIGHT = 0.50;
+constexpr double LAST_HOUR_PENALTY_WEIGHT = 0.25;
+constexpr double MORE_THAN_HOUR_PENALTY = 0.05;
+
 } // namespace
 
+template <class Params>
 struct KmsUrlCtx {
+	enum class PenaltyType { TIMEOUT = 1, MALFORMED_RESPONSE = 2 };
+
 	std::string url;
 	uint64_t nRequests;
 	uint64_t nFailedResponses;
 	uint64_t nResponseParseFailures;
+	double unresponsivenessPenalty;
+	int64_t unresponsivenessPenaltyTS;
 
-	KmsUrlCtx() : url(""), nRequests(0), nFailedResponses(0), nResponseParseFailures(0) {}
-	explicit KmsUrlCtx(const std::string& u) : url(u), nRequests(0), nFailedResponses(0), nResponseParseFailures(0) {}
+	KmsUrlCtx()
+	  : url(""), nRequests(0), nFailedResponses(0), nResponseParseFailures(0), unresponsivenessPenalty(0.0),
+	    unresponsivenessPenaltyTS(0) {}
+	explicit KmsUrlCtx(const std::string& u)
+	  : url(u), nRequests(0), nFailedResponses(0), nResponseParseFailures(0), unresponsivenessPenalty(0.0),
+	    unresponsivenessPenaltyTS(0) {}
 
-	bool operator<(const KmsUrlCtx& toCompare) const {
-		if (nFailedResponses != toCompare.nFailedResponses) {
-			return nFailedResponses > toCompare.nFailedResponses;
+	bool operator==(const KmsUrlCtx& toCompare) const { return url.compare(toCompare.url) == 0; }
+
+	void refreshUnresponsivenessPenalty() {
+		if (unresponsivenessPenaltyTS == 0) {
+			return;
+		}
+		int64_t timeSinceLastPenalty = now() - unresponsivenessPenaltyTS;
+		unresponsivenessPenalty = Params::penalty(timeSinceLastPenalty);
+	}
+
+	void penalize(const PenaltyType type) {
+		if (type == PenaltyType::TIMEOUT) {
+			nFailedResponses++;
+			unresponsivenessPenaltyTS = now();
+		} else {
+			ASSERT_EQ(type, PenaltyType::MALFORMED_RESPONSE);
+			nResponseParseFailures++;
 		}
-		return nResponseParseFailures > toCompare.nResponseParseFailures;
+	}
+
+	std::string toString() const {
+		return fmt::format(
+		    "{} {} {} {} {}", url, nRequests, nFailedResponses, nResponseParseFailures, unresponsivenessPenalty);
```

Comment:
nit: do we want some text here to know what is what? Otherwise we have to always refer to the code

Reply:
`toString` are used with an object type, so not needed.